File: engine/src/theseus_engine/queries.py
"""Spatial utility queries — small solvers over the shared world model.

The payoff of the "one world model, many solvers" rule is that features
like these cost ~20 lines each, because the grid already maintains
clearance, semantics and costs:

- corridor_profile / fits_through — "will this couch fit through there?"
  v0 is disk-swept: a route admits width w iff every point along it has
  clearance >= w/2. (An oriented-rectangle sweep is the M6 refinement;
  the disk model is conservative for long objects cornering and slightly
  permissive for diagonal shimmies — honest enough to flag pinch points.)
- nearest_semantic — "take me to the closest <label>": a flow field whose
  goals are every traversable cell ADJACENT to a matching labeled cell
  (you stand next to the fridge, not inside it).
"""

from __future__ import annotations

from .astar import PlanResult
from .flowfield import FlowField
from .geometry import INF, Vec, dist, point_along
from .grid import OccupancyGrid, PlanParams


def corridor_profile(grid: OccupancyGrid, pts: list[Vec],
                     sample_m: float = 0.15) -> list[tuple[float, float]]:
    """(arc_length, corridor_width) samples along a world-space polyline.
    Corridor width at a point is twice the clearance field there."""
    if len(pts) < 2:
        c = grid.world_to_cell(pts[0]) if pts else (0, 0)
        return [(0.0, 2.0 * grid.clearance(c))]
    out: list[tuple[float, float]] = []
    s = 0.0
    for i in range(len(pts) - 1):
        a, b = pts[i], pts[i + 1]
        seg = dist(a, b)
        steps = max(1, int(seg / sample_m))
        last = steps + 1 if i == len(pts) - 2 else steps  # include endpoint once
        for k in range(last):
            t = k / steps
            p = (a[0] + (b[0] - a[0]) * t, a[1] + (b[1] - a[1]) * t)
            width = 2.0 * grid.clearance(grid.world_to_cell(p))
            out.append((s + seg * t, width))
        s += seg
    return out


def fits_through(grid: OccupancyGrid, pts: list[Vec], width_m: float,
                 margin_m: float = 0.05) -> tuple[bool, Vec | None, float]:
    """Can an object of the given width ride this route?
    Returns (verdict, pinch_point_or_None, narrowest_width)."""
    need = width_m + 2.0 * margin_m
    worst_s, worst_w = 0.0, INF
    for s, w in corridor_profile(grid, pts):
        if w < worst_w:
            worst_s, worst_w = s, w
    if worst_w >= need:
        return True, None, worst_w
    pinch = point_along(pts, 0, 0.0, worst_s) if len(pts) >= 2 else pts[0]
    return False, pinch, worst_w


def nearest_semantic(grid: OccupancyGrid, start, label: str,
                     params: PlanParams,
                     approach_m: float = 0.75) -> PlanResult | None:
    """Optimal route to the closest object carrying this semantic label,
    or None. The route ends at an *approach point*: the nearest cell the
    BODY can legally occupy within approach_m (arm's reach) of the
    object. Strictly adjacent cells never qualify — one cell away from
    furniture is ~6 cm of clearance, and nobody's radius fits there;
    "at the fridge" means standing in front of it, not inside it."""
    labeled = [(i % grid.width, i // grid.width)
               for i, lab in grid.labels.items() if lab == label]
    if not labeled:
        return None
    r = max(1, int(approach_m / grid.cell_size))
    near: set = set()
    for (x, y) in labeled:
        for dy in range(-r, r + 1):
            for dx in range(-r, r + 1):
                if dx * dx + dy * dy <= r * r:
                    near.add((x + dx, y + dy))
    goals = [c for c in near
             if grid.in_bounds(c) and grid.traversable(c, params)]
    if not goals:
        return None
    return FlowField.build(grid, sorted(goals), params).route_from(start)

File: apps/navcore/Sources/NavCore/Queries.swift
// Spatial utility queries — port of engine queries.py. The payoff of
// "one world model, many solvers": each of these costs ~20 lines
// because the grid already maintains clearance, semantics and costs.

/// (arcLength, corridorWidth) samples along a world-space polyline.
/// Corridor width at a point is twice the clearance field there.
public func corridorProfile(grid: OccupancyGrid, pts: [Vec],
                            sampleM: Double = 0.15) -> [(Double, Double)] {
    if pts.count < 2 {
        let c = pts.isEmpty ? Cell(0, 0) : grid.worldToCell(pts[0])
        return [(0.0, 2.0 * grid.clearance(c))]
    }
    var out: [(Double, Double)] = []
    var s = 0.0
    for i in 0..<(pts.count - 1) {
        let a = pts[i]
        let b = pts[i + 1]
        let seg = dist(a, b)
        let steps = max(1, Int(seg / sampleM))
        let last = i == pts.count - 2 ? steps + 1 : steps  // endpoint once
        for k in 0..<last {
            let t = Double(k) / Double(steps)
            let p = Vec(a.x + (b.x - a.x) * t, a.y + (b.y - a.y) * t)
            let width = 2.0 * grid.clearance(grid.worldToCell(p))
            out.append((s + seg * t, width))
        }
        s += seg
    }
    return out
}

/// Can an object of the given width ride this route?
/// Returns (verdict, pinch point or nil, narrowest width).
public func fitsThrough(grid: OccupancyGrid, pts: [Vec], widthM: Double,
                        marginM: Double = 0.05) -> (Bool, Vec?, Double) {
    let need = widthM + 2.0 * marginM
    var worstS = 0.0
    var worstW = INF
    for (s, w) in corridorProfile(grid: grid, pts: pts) {
        if w < worstW {
            worstS = s
            worstW = w
        }
    }
    if worstW >= need {
        return (true, nil, worstW)
    }
    let pinch = pts.count >= 2
        ? pointAlong(pts, startI: 0, startT: 0.0, ahead: worstS)
        : pts[0]
    return (false, pinch, worstW)
}

/// Optimal route to the closest object carrying this semantic label, or
/// nil. The route ends at an *approach point*: the nearest cell the BODY
/// can legally occupy within approachM (arm's reach) of the object —
/// "at the fridge" means standing in front of it, not inside it.
public func nearestSemantic(grid: OccupancyGrid, start: Cell, label: String,
                            params: PlanParams,
                            approachM: Double = 0.75) -> PlanResult? {
    let labeled = grid.labels.compactMap { (i, lab) -> Cell? in
        lab == label ? Cell(Int32(i % grid.width), Int32(i / grid.width))
                     : nil
    }
    if labeled.isEmpty {
        return nil
    }
    let r = max(1, Int(approachM / grid.cellSize))
    var near = Set<Cell>()
    for c in labeled {
        for dy in -r...r {
            for dx in -r...r {
                if dx * dx + dy * dy <= r * r {
                    near.insert(Cell(c.x + Int32(dx), c.y + Int32(dy)))
                }
            }
        }
    }
    let goals = near.filter { grid.inBounds($0) && grid.traversable($0, params) }
        .sorted { ($0.x, $0.y) < ($1.x, $1.y) }
    if goals.isEmpty {
        return nil
    }
    return FlowField.build(grid: grid, goals: goals, params: params)
        .routeFrom(start)
}
